lib/ClangImporter/SwiftLookupTable.h

```diff
@@ -45,6 +45,102 @@ class TypedefNameDecl;
 
 namespace swift {
 
+/// A name from a SwiftLookupTable that has not been deserialized into a
+/// DeclBaseName yet
+struct SerializedSwiftName {
+  /// The kind of the name if it is a special name
+  DeclBaseName::Kind Kind;
+  /// The name of the idenifier if it is not a special name
+  StringRef Name;
+
+  SerializedSwiftName() : Kind(DeclBaseName::Kind::Normal), Name(StringRef()) {}
+
+  explicit SerializedSwiftName(DeclBaseName::Kind Kind)
+      : Kind(Kind), Name(StringRef()) {}
+
+  explicit SerializedSwiftName(StringRef Name)
+      : Kind(DeclBaseName::Kind::Normal), Name(Name) {}
+
+  SerializedSwiftName(DeclBaseName BaseName) {
+    Kind = BaseName.getKind();
+    if (BaseName.getKind() == DeclBaseName::Kind::Normal) {
+      Name = BaseName.getIdentifier().str();
+    }
+  }
+
+  /// Deserialize the name, adding it to the context's identifier table
+  DeclBaseName toDeclBaseName(ASTContext &Context) const;
+
+  bool empty() const {
+    return Kind == DeclBaseName::Kind::Normal && Name.empty();
+  }
+
+  /// Return a string representation of the name that can be used for sorting
+  StringRef sortStr() const {
+    switch (Kind) {
+    case DeclBaseName::Kind::Normal:
+      return Name;
+    case DeclBaseName::Kind::Subscript:
+      return "subscript";
+    }
+  }
+
+  bool operator<(SerializedSwiftName RHS) const {
+    return sortStr() < RHS.sortStr();
+  }
+
+  bool operator==(SerializedSwiftName RHS) const {
+    if (Kind != RHS.Kind)
+      return false;
+
+    if (Kind == DeclBaseName::Kind::Normal) {
+      assert(RHS.Kind == DeclBaseName::Kind::Normal);
+      return Name == RHS.Name;
+    } else {
+      return true;
+    }
+  }
+};
+
+} // end namespace swift
+
+namespace llvm {
+
+using swift::SerializedSwiftName;
+
+// Inherit the DenseMapInfo from StringRef but add a few special cases for
+// special names
+template<> struct DenseMapInfo<SerializedSwiftName> {
+  static SerializedSwiftName getEmptyKey() {
+    return SerializedSwiftName(DenseMapInfo<StringRef>::getEmptyKey());
+  }
+  static SerializedSwiftName getTombstoneKey() {
+    return SerializedSwiftName(DenseMapInfo<StringRef>::getTombstoneKey());
+  }
+  static unsigned getHashValue(SerializedSwiftName Val) {
+    if (Val.Kind == swift::DeclBaseName::Kind::Normal) {
+      return DenseMapInfo<StringRef>::getHashValue(Val.Name);
+    } else {
+      return (unsigned)Val.Kind;
+    }
+  }
+  static bool isEqual(SerializedSwiftName LHS, SerializedSwiftName RHS) {
```

Comment:
It has a separate implementation from `==` in order to call through to `DenseMapInfo<StringRef>::getHashValue` to get correct correct comparisons for the `StringRef` tombstone and empty key. I guess calling `==` doesn't work because it uses `StringRef`'s `==` operator and since both the `StringRef` tombstone and empty key have length 0 they would end up being equal, which is incorrect. Didn't check though.